packages/replay-next/src/suspense/createGenericRangeCache.ts

```diff
@@ -0,0 +1,227 @@
+
+import {
+  Range,
+  getMissingRanges,
+  isInRange,
+  mergeRanges,
+  rangeContains,
+  rangesOverlap,
+} from "../utils/range-list";
+import { Subscriber, UnsubscribeFunction } from "./SearchCache";
+import { Thennable } from "./types";
+
+export interface RangeCache<T> {
+  getValuesSuspense(client: ReplayClientInterface, range: PointRange): T[];
+  getValuesAsync(client: ReplayClientInterface, range: PointRange): Thennable<T[]> | T[];
+  getCachedValues(range: PointRange): T[];
+}
+
+interface RangeStore<T> {
+  cachedRanges: Range[];
+  valuesFromCachedRanges: T[];
+  pendingRanges: Range[];
+  valuesFromPendingRanges: T[];
+  subscribe(subscriber: Subscriber): UnsubscribeFunction;
+  notifySubscribers(): void;
+  fetchMissingRanges(client: ReplayClientInterface, range: PointRange): Promise<void>;
+  getExecutionPoint: (value: T) => ExecutionPoint;
+}
+
+interface PendingRange<T> {
+  range: Range;
+  values: T[];
+  promise?: Promise<void>;
+}
+
+interface FailedRange {
+  range: Range;
+  error: any;
+}
+
+export function createGenericRangeCache<T>(
+  fetchRange: (
+    client: ReplayClientInterface,
+    range: PointRange,
+    cacheValues: (values: T[]) => void,
+    cacheError: (error: any) => void
+  ) => Promise<void>,
+  getExecutionPoint: (value: T) => ExecutionPoint
+): RangeCache<T> {
+  const pendingRanges: PendingRange<T>[] = [];
+  const failedRanges: FailedRange[] = [];
+  const subscribers: Set<Subscriber> = new Set();
+  const store: RangeStore<T> = {
+    cachedRanges: [],
+    valuesFromCachedRanges: [],
+    pendingRanges: [],
+    valuesFromPendingRanges: [],
+    subscribe(subscriber: Subscriber) {
+      subscribers.add(subscriber);
+      return () => {
+        subscribers.delete(subscriber);
+      };
+    },
+    notifySubscribers() {
+      subscribers.forEach(subscriber => subscriber());
+    },
+    async fetchMissingRanges(client: ReplayClientInterface, pointRange: PointRange) {
+      const requestedRange = pointRangeToRange(pointRange);
+      const missingRanges = getMissingRanges(
+        requestedRange,
+        mergeRanges(store.cachedRanges, store.pendingRanges)
+      );
+
+      const failedRange = failedRanges.find(failedRange =>
+        missingRanges.some(missingRange => rangeContains(missingRange, failedRange.range))
+      );
+      if (failedRange) {
+        throw failedRange.error;
+      }
+
+      missingRanges.forEach(missingRange => {
+        const pendingRange: PendingRange<T> = {
+          range: missingRange,
+          values: [],
+        };
+
+        pendingRange.promise = new Promise<void>(async (resolve, reject) => {
+          pendingRanges.push(pendingRange);
+          store.pendingRanges = mergeRanges(store.pendingRanges, [missingRange]);
+
+          let failed = false;
+          function onError(error: any) {
+            if (!failed) {
+              failed = true;
+              failedRanges.push({ range: missingRange, error });
```

Comment:
Maybe I'm reading this wrong, but I _think_ we should only gate the call to `reject` in `if (!failed)` but we should always push failed ranges?

These generic caches should have unit tests too I think, that explicitly cover edge cases like this.

Reply:
If `failed` is `true` then we have already pushed the `missingRange` to `failedRanges`, no need to do that again.

Reply:
That's only true for a single range though, right? Can't there be more than one?

Reply:
Yes, there can be more than one, but since `failed` is declared within the loop for the ranges, each range gets its own `failed` flag.

Reply:
I see.